### DISCUSSION TITLE
Making Datasette easier to install

### DISCUSSION THREAD

**Discussion by simonw** [simonw/datasette#921] [General]
Anyone have any thoughts on how I can make Datasette easier to install?

Right now the instructions at https://datasette.readthedocs.io/en/stable/installation.html talk about pip, pipx and Docker. But people need to get a Python environment working, which is really tricky.

I want to get Homebrew working too: https://github.com/simonw/datasette/issues/335

**Comment by obra** [comment]
Maybe consider having "one preferred way" in the intro instructions?

pipx feels...advanced. Maybe break it out into its own page to make the installation seem easier?

**Reply by simonw** [reply]
I'm really struggling with this at the moment.

I guess teaching people to install it globally (or ideally using `pip install datasette --user`) avoids me introducing a new tool... but Datasette definitely benefits from having its own environment since there's so much scope for plugins and extra dependencies.

The Homebrew version will ensure Datasette gets its own virtual environment without the user having to think about it, which is good. Only serves macOS users though.

I need to spend time thinking about Windows too, but I'm really not very familiar with what works well in that environment.

**Reply by simonw** [reply]
But yes, I like the idea of splitting out the pipx stuff into an "advanced" section. The more I can focus on simplicity-of-installation the better.

**Reply by obra** [reply]
There are two kinds of installs, I think. "intfrastructurey" installs and "casual" installs. And "infrastructurey" installs should absolutely have an isolated environment. But casual installs will be most users' first experience with Datasette. And for those, I think you want to optimize for "few commands, less time, fewer deps"

Honestly, for the well-encapsulated infrastructurey installations, maybe you just always want to push people at Docker as the supported environment? It works on Windows, macOS and Linux. You have a known-good environment. And you can do a little bit of light tooling around it to make it simple to interact with.

**Reply by simonw** [reply]
That's a useful distinction - I like "infrastructurey" v.s. "casual".

How likely are people to have Docker for Desktop installed though? I worry that it's a pretty enormous dependency to drop on people.

**Reply by obra** [reply]
Well, you don't force it on folks. 

To me, it's a "cross-platform package"

But you have limited support resources and can't (currently) support every possible platform and packaging option. So having one good choice that works everywhere and is well tested and well docced isn't wrong.

**Reply by rayvoelker** [reply]
Maybe I'm bias, but I still feel like the "easy" platform for anything Python-related is Linux. There are plenty of ways to get either a VM / VPS / Docker (or even running locally ... really, why are you still running windows?) and then working within Python's now-native venv module within most modern Linux distros.

https://www.python.org/dev/peps/pep-0405/

**Reply by simonw** [reply]
That's probably true for people who are going to be writing code, but my target audience for Datasette is more people who want to do some data analysis on their laptop.

That's why I want to stay away from talking them through a Docker install or running a virtual machine if I can possibly avoid it.

**Reply by rayvoelker** [reply]
I very much appreciate all the thought and effort you're putting in to make this software as user friendly as possible. I had had the thought of trying to leverage Datasette in providing a platform on which to teach some (just-a-little-bit technical) folks from public and private libraries how to use some SQL features of the catalog system that many of us use.

The read-only, web-based nature of Datasette seems perfect for such a task.... and the easier it is to deploy, the easier folks can practice their skills further!

**Comment by simonw** [comment]
A longer-term ambition I have is to bundle up Datasette as an Electron app - primarily for ease of installation by users who aren't comfortable with the terminal.

**Reply by rcoup** [reply]
@simonw for casual users (particularly Mac & Windows without Homebrew/Chocolatey), there's always [PyInstaller](https://www.pyinstaller.org/) — bundle the interpreter and all with it, then wrap it in a native OS installer. For Linux deb/rpm packaging with the correct dependencies on Python/etc is probably simpler.

**Comment by phildini** [comment]
I'm obviously biased here, but you might look at https://github.com/phildini/hera

Hera is a standalone mac app that runs a jupyter notebook server, and even supports putting `pip install` -ed plugins in the right place.

It's using BeeWare's briefcase, so it's all python.

Update: here's a mac app that runs datasette, written in python: https://github.com/phildini/walkman

**Reply by simonw** [reply]
Oh that is really interesting! I was talking to Asheesh about how hard it would be to run Datasette on Android using BeeWare and it definitely sounded feasible. I would LOVE to be able to run Datasette on phones - feels like a much more secure way for people to browse their own personal data (see [Dogsheep](https://dogsheep.github.io/)) than uploading it to the cloud somewhere.

**Reply by phildini** [reply]
@simonw here's a mac app that runs datasette. https://github.com/phildini/walkman

**Reply by simonw** [reply]
Whoa! That totally worked (I had to kill the datasette process I already had running on port 8001 first though):

<img width="785" alt="Screenshot_8_11_20__2_39_PM" src="https://user-images.githubusercontent.com/9599/89951611-978d5280-dbe0-11ea-8429-713eec6f283e.png">

How hard is it to package that up into a Mac app people can install without using pipenv and pip?

**Comment by obra** [comment]
Minor thing:

You've got at least one note in the docs about things like "how to install on an old python rev (3.5)". I don't think that needs to be in your intro docs.

**Reply by simonw** [reply]
Yeah I'll ditch that, I think we can all agree it's time.

**Reply by simonw** [reply]
... done https://github.com/simonw/datasette/commit/3fa261d1d2034b9444526ad764600f9798d9058c

**Comment by jcmkk3** [comment]
For Windows, I most prefer to install applications (especially cli) using [Scoop](https://github.com/lukesampson/scoop). The user does not need to be an admin to install through Scoop, which is a huge plus on corporate machines. Microsoft is working on their own package manager  ([winget](https://github.com/microsoft/winget-cli)) which might be worth looking at, as well.

**Reply by simonw** [reply]
That's really useful, thanks!

**Reply by dracos** [reply]
Conversely, on Windows I have WSL installed and just use that for everything :)

**Reply by abdusco** [reply]
there's also an alternative package repo for Windows called Chocolatey https://chocolatey.org/

**Reply by pfmoore** [reply]
Bundling a Python application with the "embedded distribution" and the application code, as a scoop application, is pretty easy in many cases. I might take a look at it - I've not used datasette for much directly (never had the time) but as a database guy, I've always been interested in it, and the problem triggers my packaging itch 🙂

**Reply by simonw** [reply]
That would be fantastic!

**Reply by pfmoore** [reply]
OK. Take a look at https://github.com/pfmoore/datasette-standalone

You can either download the release files directly and use them, or you can use scoop as described in the project readme. There are still some rough edges here:

1. The release process is manual at the moment, so this won't track datasette releases unless I remember to do an update. This is something I want to automate, but I decided to get something out and worry about that later.
2. Using it without scoop, you have to run `python -m datasette` using the Python that's in the archive. That's clumsy. I'd like to provide a `datasette.exe` binary, but that's a little bit more work, and I don't know whether there's enough interest to warrant it. The scoop experience is much better - you get a `datasette` command directly.
3. This may be the big one. There's no way of adding plugins unless you know what you're doing and are happy editing the installation directly with pip. The thing is, I don't really see how it's practical to manage plugins *without* using pip, and that seems somewhat counter to the idea of having a standalone application `datasette`. So I don't really know what you'd want or expect here. For now I'm leaving this as "not supported unless you know what you're doing".

Hope this is useful - suggestions are welcome, either here or on the project tracker!

**Reply by simonw** [reply]
Have you tried the new `datasette install datasette-Vega` command from Datasette 0.47?

It's a very thin wrapper around `pip install` which exists just so users of things like Homebrew don't have to figure out where the right pip is!

**Reply by pfmoore** [reply]
No, I hadn't - that's cool. It doesn't work at the moment because I don't include pip in the build, but I can fix that. I'll need to do some testing to see whether scoop is OK with users installing "extra stuff" too. Leave it with me :-)

**Reply by pfmoore** [reply]
Right, I have an updated version now. Installing extensions seems to work from my experiments. The zipped release files now include a `datasette.exe` that can be used to run datasette, for anyone who doesn't want to use the scoop version.

**Reply by pfmoore** [reply]
I just updated the release to datasette 0.48

**Comment by simonw** [comment]
I *almost* got homebrew working!

```
$ brew tap simonw/datasette
$ brew install datasette
$ /usr/local/Cellar/datasette/0.46/libexec/bin/datasette --version
datasette, version 0.46
```
There's just one catch:
```
$ which datasette
datasette not found
```
So it's installed, but it wasn't correctly symlinked from somewhere on the PATH (usually Homebrew appears to symlink from `/usr/local/bin`). And I have no idea why.

Here's an issue if anyone has any ideas: https://github.com/simonw/homebrew-datasette/issues/2

**Reply by simonw** [reply]
Got it working! You can now install Datasette using:

    brew install simonw/datasette/datasette

**Reply by simonw** [reply]
Over night I got the formula accepted into Homebrew core, which means you can now run `brew install datasette`

**Comment by simonw** [comment]
Datasette 0.47 is out with a bunch of changes to try to make it easier to install, and a hopefully much improved installation page in the documentation: https://datasette.readthedocs.io/en/stable/installation.html

**Comment by cfljam** [comment]
```
conda create -n my_datasette_env  python=3.6
pip install datasette 
datasette my.db
```

works for me..such a wonderful tool for so little work. Thanks Simon
